File: zenscript-code-model/src/main/java/raylras/zen/model/type/MemberValidator.java
package raylras.zen.model.type;

import raylras.zen.model.symbol.*;

import java.util.*;

public class MemberValidator implements Iterable<Symbol> {
    private final Map<String, Symbol> fields = new HashMap<>();
    private final Map<MemberValidator.ExecutableData, Symbol> executables = new HashMap<>();
    private OperatorFunctionSymbol caster = null;

    public void add(Symbol symbol) {
        if (symbol instanceof Executable executable) {
            if (symbol instanceof OperatorFunctionSymbol operator && operator.getOperator() == Operator.AS) {
                addCaster(operator);
            } else {
                addExecutable((Symbol & Executable) executable);
            }
        } else {
            addField(symbol);
        }
    }

    public void addAll(Iterable<? extends Symbol> other) {
        other.forEach(this::add);
    }

    public Collection<Symbol> getMembers() {
        List<Symbol> symbols = new ArrayList<>(fields.size() + executables.size() + 1);
        symbols.addAll(fields.values());
        symbols.addAll(executables.values());
        if (caster != null) {
            symbols.add(caster);
        }
        return List.copyOf(symbols);
    }

    private void addField(Symbol symbol) {
        fields.putIfAbsent(symbol.getName(), symbol);
    }

    private <T extends Symbol & Executable> void addExecutable(T executableSymbol) {
        List<Type> parameterTypes = executableSymbol.getParameterList().stream().map(Symbol::getType).toList();
        executables.putIfAbsent(new MemberValidator.ExecutableData(executableSymbol.getName(), executableSymbol.getKind(), parameterTypes), executableSymbol);
    }

    private void addCaster(OperatorFunctionSymbol operatorFunctionSymbol) {
        if (caster == null) {
            caster = operatorFunctionSymbol;
        } else {
            caster = SymbolFactory.createOperatorFunctionSymbol(
                    Operator.AS,
                    new IntersectionType(List.of(caster.getReturnType(), operatorFunctionSymbol.getReturnType())),
                    Collections.emptyList()
            );
        }
    }

    @Override
    public Iterator<Symbol> iterator() {
        return getMembers().iterator();
    }

    private record ExecutableData(String name, Symbol.Kind kind, List<Type> parameters) {}
}

File: zenscript-code-model/src/main/kotlin/raylras/zen/model/type/MemberValidator.kt
package raylras.zen.model.type

import raylras.zen.model.symbol.Executable
import raylras.zen.model.symbol.Operator
import raylras.zen.model.symbol.OperatorFunctionSymbol
import raylras.zen.model.symbol.Symbol
import raylras.zen.model.symbol.impl.createOperatorFunctionSymbol

class MemberValidator {
    private val properties = HashMap<String, Symbol>()
    private val executables = HashMap<ExecutableData, Symbol>()
    private val casterTypes = ArrayList<Type>()

    fun add(symbol: Symbol) {
        when (symbol) {
            is Executable -> {
                when {
                    symbol is OperatorFunctionSymbol && symbol.operator == Operator.AS -> {
                        addCaster(symbol)
                    }
                    else -> {
                        addExecutable(symbol)
                    }
                }
            }

            else -> {
                addProperty(symbol)
            }
        }
    }

    fun getMembers(): Collection<Symbol> {
        val members = ArrayList<Symbol>(properties.size + executables.size + 1)
        members.addAll(properties.values)
        members.addAll(executables.values)
        when {
            casterTypes.size == 1 -> {
                members.add(createOperatorFunctionSymbol(Operator.AS, casterTypes[0]))
            }
            casterTypes.size > 1 -> {
                members.add(createOperatorFunctionSymbol(Operator.AS, IntersectionType(casterTypes)))
            }
        }
        return members
    }

    private fun addProperty(symbol: Symbol) {
        properties.putIfAbsent(symbol.simpleName, symbol)
    }

    private fun <T> addExecutable(symbol: T) where T : Symbol, T : Executable {
        val parameterTypes = symbol.parameters.map { it.type }.toList()
        executables.putIfAbsent(
            ExecutableData(symbol.simpleName, parameterTypes),
            symbol
        )
    }

    private fun addCaster(caster: OperatorFunctionSymbol) {
        when (val returnType = caster.returnType) {
            is IntersectionType -> {
                casterTypes.addAll(returnType.typeList)
            }

            else -> {
                casterTypes.add(returnType)
            }
        }
    }

    private data class ExecutableData(val name: String, val parameters: List<Type>)
}
